[near-cache] MemberMapReconciliationTest

**com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.test_reconciliation_does_not_cause_premature_removal**
```java
java.lang.AssertionError: not expected expectedHits expected:<100> but was:<99>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.assertStats(MemberMapReconciliationTest.java:118)
	at com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.test_reconciliation_does_not_cause_premature_removal(MemberMapReconciliationTest.java:113)
```
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/10401/testReport/junit/com.hazelcast.map.impl.nearcache.invalidation/MemberMapReconciliationTest/test_reconciliation_does_not_cause_premature_removal/

Should be the same pattern as in the other Near Cache tests -> invalidations from initial population being processed late on the Near Cache instance.

***com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.test_reconciliation_does_not_cause_premature_removal[mapInMemoryFormat:BINARY nearCacheInMemoryFormat:OBJECT]***

https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/11033

```
java.lang.AssertionError: Wrong expectedHits [NearCacheStatsImpl{ownedEntryCount=100, ownedEntryMemoryCost=0, creationTime=1506960992132, hits=99, misses=101, ratio=98.0%, evictions=0, expirations=0, lastPersistenceTime=0, persistenceCount=0, lastPersistenceDuration=0, lastPersistenceWrittenBytes=0, lastPersistenceKeyCount=0, lastPersistenceFailure=''}] expected:<100> but was:<99>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.assertStats(MemberMapReconciliationTest.java:155)
	at com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.test_reconciliation_does_not_cause_premature_removal(MemberMapReconciliationTest.java:149)
```

Also failed in https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/11754:
```
java.lang.AssertionError: Wrong expectedHits [NearCacheStatsImpl{ownedEntryCount=100, ownedEntryMemoryCost=10800, creationTime=1511337876276, hits=99, misses=101, ratio=98.0%, evictions=0, expirations=0, lastPersistenceTime=0, persistenceCount=0, lastPersistenceDuration=0, lastPersistenceWrittenBytes=0, lastPersistenceKeyCount=0, lastPersistenceFailure=''}] expected:<100> but was:<99>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.assertStats(MemberMapReconciliationTest.java:155)
	at com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.test_reconciliation_does_not_cause_premature_removal(MemberMapReconciliationTest.java:149)
```

**com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.test_reconciliation_does_not_cause_premature_removal[mapInMemoryFormat:OBJECT nearCacheInMemoryFormat:OBJECT]**
```java
java.lang.AssertionError: Wrong expectedHits [NearCacheStatsImpl{ownedEntryCount=100, ownedEntryMemoryCost=0, creationTime=1511974070623, hits=96, misses=104, ratio=92.3%, evictions=0, expirations=0, lastPersistenceTime=0, persistenceCount=0, lastPersistenceDuration=0, lastPersistenceWrittenBytes=0, lastPersistenceKeyCount=0, lastPersistenceFailure=''}] expected:<100> but was:<96>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.assertStats(MemberMapReconciliationTest.java:155)
	at com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.test_reconciliation_does_not_cause_premature_removal(MemberMapReconciliationTest.java:149)
```
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/11910/testReport/junit/com.hazelcast.map.impl.nearcache.invalidation/MemberMapReconciliationTest/test_reconciliation_does_not_cause_premature_removal_mapInMemoryFormat_OBJECT_nearCacheInMemoryFormat_OBJECT_/

**com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.test_reconciliation_does_not_cause_premature_removal[mapInMemoryFormat:OBJECT nearCacheInMemoryFormat:OBJECT]**
```java
java.lang.AssertionError: Wrong expectedHits [NearCacheStatsImpl{ownedEntryCount=100, ownedEntryMemoryCost=0, creationTime=1519848266317, hits=98, misses=102, ratio=96.1%, evictions=0, expirations=0, lastPersistenceTime=0, persistenceCount=0, lastPersistenceDuration=0, lastPersistenceWrittenBytes=0, lastPersistenceKeyCount=0, lastPersistenceFailure=''}] expected:<100> but was:<98>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.assertStats(MemberMapReconciliationTest.java:155)
	at com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.test_reconciliation_does_not_cause_premature_removal(MemberMapReconciliationTest.java:149)
```
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/13960/testReport/junit/com.hazelcast.map.impl.nearcache.invalidation/MemberMapReconciliationTest/test_reconciliation_does_not_cause_premature_removal_mapInMemoryFormat_OBJECT_nearCacheInMemoryFormat_OBJECT_/

Failure occurred again in the PR builder https://hazelcast-l337.ci.cloudbees.com/view/Lab/job/new-lab-fast-pr/14548/testReport/junit/com.hazelcast.map.impl.nearcache.invalidation/MemberMapReconciliationTest/test_reconciliation_does_not_cause_premature_removal_mapInMemoryFormat_OBJECT_nearCacheInMemoryFormat_OBJECT_/

Stacktrace:
```
java.lang.AssertionError: Wrong expectedHits [NearCacheStatsImpl{ownedEntryCount=100, ownedEntryMemoryCost=0, creationTime=1521549446094, hits=99, misses=101, ratio=98.0%, evictions=0, expirations=0, lastPersistenceTime=0, persistenceCount=0, lastPersistenceDuration=0, lastPersistenceWrittenBytes=0, lastPersistenceKeyCount=0, lastPersistenceFailure=''}] expected:<100> but was:<99>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.assertStats(MemberMapReconciliationTest.java:155)
	at com.hazelcast.map.impl.nearcache.invalidation.MemberMapReconciliationTest.test_reconciliation_does_not_cause_premature_removal(MemberMapReconciliationTest.java:149)
```

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/nearcache/invalidation/MemberMapReconciliationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/nearcache/invalidation/MemberMapReconciliationTest.java
@@ -22,8 +22,13 @@ import com.hazelcast.config.MapConfig;
 import com.hazelcast.config.NearCacheConfig;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.IMap;
+import com.hazelcast.core.Partition;
+import com.hazelcast.internal.nearcache.impl.DefaultNearCache;
+import com.hazelcast.internal.nearcache.impl.invalidation.MetaDataContainer;
+import com.hazelcast.internal.nearcache.impl.invalidation.StaleReadDetector;
 import com.hazelcast.map.impl.proxy.NearCachedMapProxyImpl;
 import com.hazelcast.monitor.NearCacheStats;
+import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastParallelParametersRunnerFactory;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
@@ -47,12 +52,15 @@ import static com.hazelcast.internal.nearcache.NearCacheTestUtils.getBaseConfig;
 import static com.hazelcast.internal.nearcache.impl.invalidation.RepairingTask.MAX_TOLERATED_MISS_COUNT;
 import static com.hazelcast.internal.nearcache.impl.invalidation.RepairingTask.MIN_RECONCILIATION_INTERVAL_SECONDS;
 import static com.hazelcast.internal.nearcache.impl.invalidation.RepairingTask.RECONCILIATION_INTERVAL_SECONDS;
+import static com.hazelcast.internal.nearcache.impl.invalidation.StaleReadDetector.ALWAYS_FRESH;
 import static com.hazelcast.spi.properties.GroupProperty.MAP_INVALIDATION_MESSAGE_BATCH_FREQUENCY_SECONDS;
 import static com.hazelcast.spi.properties.GroupProperty.MAP_INVALIDATION_MESSAGE_BATCH_SIZE;
 import static java.lang.String.format;
 import static java.lang.String.valueOf;
 import static java.util.Arrays.asList;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNotSame;
 
 @RunWith(Parameterized.class)
 @UseParametersRunnerFactory(HazelcastParallelParametersRunnerFactory.class)
@@ -83,6 +91,7 @@ public class MemberMapReconciliationTest extends HazelcastTestSupport {
 
     private Config config;
 
+    private HazelcastInstance server;
     private IMap<Integer, Integer> serverMap;
     private IMap<Integer, Integer> nearCachedServerMap;
 
@@ -107,7 +116,7 @@ public class MemberMapReconciliationTest extends HazelcastTestSupport {
                 .setProperty(MAP_INVALIDATION_MESSAGE_BATCH_SIZE.getName(), valueOf(Integer.MAX_VALUE))
                 .addMapConfig(mapConfig);
 
-        HazelcastInstance server = factory.newHazelcastInstance(config);
+        server = factory.newHazelcastInstance(config);
         serverMap = server.getMap(MAP_NAME);
         nearCachedServerMap = nearCachedMapFromNewServer();
     }
@@ -128,10 +137,11 @@ public class MemberMapReconciliationTest extends HazelcastTestSupport {
             nearCachedServerMap.get(i);
         }
 
-        IMap<Integer, Integer> nearCachedMapFromNewServer = nearCachedMapFromNewServer();
+        final IMap<Integer, Integer> nearCachedMapFromNewServer = nearCachedMapFromNewServer();
 
         warmUpPartitions(factory.getAllHazelcastInstances());
         waitAllForSafeState(factory.getAllHazelcastInstances());
+        waitForNearCacheInvalidationMetadata(nearCachedMapFromNewServer, server);
 
         for (int i = 0; i < total; i++) {
             nearCachedMapFromNewServer.get(i);
@@ -161,6 +171,34 @@ public class MemberMapReconciliationTest extends HazelcastTestSupport {
         assertInstanceOf(NearCachedMapProxyImpl.class, map);
 
         return map;
+    }
+
+    private static void waitForNearCacheInvalidationMetadata(final IMap<Integer, Integer> nearCachedMapFromNewServer,
+                                                             final HazelcastInstance server) {
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                final DefaultNearCache nearCache = getNearCache((NearCachedMapProxyImpl) nearCachedMapFromNewServer);
+
+                StaleReadDetector staleReadDetector = nearCache.getNearCacheRecordStore().getStaleReadDetector();
+
+                // we first assert that the stale detector is not the initial one, since the metadata that the records are
+                // initialized with on putting records into the record store is queried from the stale detector
+                assertNotSame(ALWAYS_FRESH, staleReadDetector);
+
+                // wait until all partition's metadata is filled properly, since creating records from on initial metadata
+                // may lead to stale reads if the metadata gets updated between record creation and stale read check
+                for (Partition partition : server.getPartitionService().getPartitions()) {
+                    MetaDataContainer metaDataContainer = staleReadDetector.getMetaDataContainer(partition.getPartitionId());
+
+                    assertNotNull(metaDataContainer.getUuid());
+                }
+            }
+        });
+    }
+
+    private static DefaultNearCache getNearCache(NearCachedMapProxyImpl nearCachedMapFromNewServer) {
+        return (DefaultNearCache) nearCachedMapFromNewServer.getNearCache().unwrap(DefaultNearCache.class);
     }
 
     public static void assertStats(NearCacheStats nearCacheStats, int ownedEntryCount, int expectedHits, int expectedMisses) {

